Add dist config docs (#194)
Oneflow-Inc/libai: docs/source/tutorials/Distributed_Configuration.md

# Distributed Configuration

> In LiBai, you can try out different parallel modes easily by updating the distributed config in [training config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/common/train.py).

Reviewer: ```suggestion
In LiBai, you can try out different parallel strategy by simply changing the distributed config in [training config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/common/train.py).
```

Reviewer: 这样如何

Author: 我觉得可以

Author: 已改~

        data_parallel_size=1,
        tensor_parallel_size=1,
        pipeline_parallel_size=1,
)
```
For example, you can set `data_parallel_size=2` which will automatically split the input data into two groups for data parallel training.

## Distributed Setting Example
Here we provide simple examples for users to understand the basic configuration of LiBai's distributed settings. LiBai's **BERT** model supports three parallelism techniques (**data parallel training**, **tensor parallel training** and **pipeline parallel training**), and here we use 1 node with 8 GPUs as an example. If you do not change any default settings, LiBai will execute **data parallel training** by default. You can try out different combinations of parallelism training techniques by updating [bert config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/bert_large_pretrain.py) as follows:
#### **Pure Data Parallel Training on 8 GPUs**

In this example, the model is replicated on 8 GPUs, and each replica handles only part of the input data during iteration.
```python
from .common.train import train
...

train.dist.data_parallel_size = 8
```

#### **Pure Tensor Parallel Training on 8 GPUs**

> In this example, the weight of the layers in the model be splitted into 8 parts for tensor parallel training on 8 GPUs.

Reviewer 2: ```suggestion
In this example, the weight of the layers in the model will be split into 8 parts for tensor parallel training on 8 GPUs.
```

Author: 已修改~

```python
from .common.train import train
...

train.dist.tensor_parallel_size = 8

> ```

Reviewer 3: 我觉得这里也要加一个说明, 也可以链接到   Write Models 的doc
用户只有用libai内置的layer才会生效.

Author: 可以的，这个我在Write Models下写吧，这儿就默认跑的模型是已经配置好流水了

Reviewer 3: 我看下面只有pipeline有这个说明, 可能会给用户造成误解,  其他的并行就不需要了
我感觉可能在最开始的时候加一下说明, 会比较清楚一些

Author: OK，那我都统一加一下

#### **Pure Pipeline Parallel Training on 8 GPUs**

> In this example, 8 GPUs will be splitted into 8 stages, different layers of the model will be put on different stages automatically for pipeline parallel training.

Reviewer 2: ```suggestion
In this example, 8 GPUs will be split into 8 stages, different layers of the model will be put on different stages automatically for pipeline parallel training.
```

Author: 已修改~

```python
from .common.train import train
...

train.dist.pipeline_parallel_size = 8
```

> **Note:** For models which have been configured with pipeline parallelism, e.g., BERT, GPT-2, T5 and VisionTransformer model in LiBai, you can simply update the distributed config to excute pipeline parallel training on these models. If you need to train your own model with pipeline parallel strategy, please refer to [Write Models]() for more details about configuring your own model with pipeline parallelism.

Reviewer: [Write Models]() 这里链接没更新

Author: 因为还没写，在Write Models那个PR下会更新~

Reviewer: 好的

> **Note:** For models which have been configured with pipeline parallelism, e.g., BERT, GPT-2, T5 and VisionTransformer model in LiBai, you can simply update the distributed config to excute pipeline parallel training on these models. If you need to train your own model with pipeline parallel strategy, please refer to [Write Models]() for more details about configuring your own model with pipeline parallelism.

Reviewer 2: ```suggestion
**Note:** For models configured with pipeline parallelism(e.g., BERT, GPT-2, T5, and ViT), you can simply update the distributed config to execute pipeline parallel training on them. If you need to train your own model with pipeline parallel strategy, please refer to [Write Models]() for more details about configuring your own model with pipeline parallelism.
```

Author: 已修改~

```

#### **Data Parallel + Pipeline Parallel for 2D Parallel Training on 8 GPUs**

In this example, 8 GPUs will be splitted into **2 groups**, each group contains **4 GPUs**, and the input data will be splitted into 2 parts by chunking in the batch dimension for data parallel training. The model is replicated between **2 data parellel groups**, and each group contains **4 stages**, different layers in the model will be put on different stages automatically for pipeline parallel training.
```python
from .common.train import train
...

train.dist.data_parallel_size = 2
train.dist.pipeline_parallel_size = 4
```

#### **Tensor Parallel + Pipeline Parallel for 2D Parallel Training on 8 GPUs**

In this example, 8 GPUs will be splitted into **4 stages**, each stage contains **2 GPUs** as a **group**. And different layers in the model will be put on different stages automatically for pipeline parallel training. The weight of the layers be put on the specific stage will be splitted into 2 parts for tensor parallel training within the group. 

```python
from .common.train import train
...

train.dist.tensor_parallel_size = 2
train.dist.pipeline_parallel_size = 4
```

> #### **Data Parallel + Tensor Parallel + Pipeline Parallel for 3D Parallel Training on 8 GPUs**

Reviewer: 上面 数据 + 流水并行，是把 gpu 按照 先数据并行分组，组内流水
这里 数据 + 模型 + 流水并行，是把 gpu 先按照 流水分组，组内按数据再分组，接着组内是模型

这里就是会根据 模型并行 导致分组顺序变化么？

Reviewer 3: 我个人的理解
比如 2卡的数据并行 + 2卡的流水并行,  其实分组应该是,   0和1 卡是数据并行, 在batch维度上做切分.  2卡和3卡上,  拿不到最原始的数据输入,  只能拿到0卡和1卡输出的feature.        也就是说  0卡和1卡上 是 `model_上`, 2卡和3卡上`model_下`,  然后在`model_上`这个part上面是数据并行的

Reviewer 3: 然后2+2+2 的  数据+模型+流水并行.     
0,1,2,3卡上是完整的`model_上`,   然后0卡和1卡数据并行, 有`model_前`的模型并行的上部分,  2卡和3卡也是数据并行, 有`model_前`的模型并行的下部分.  
4,5,6,7卡上是完整的`model_下`,  和上面情况是一样的, 只不过输入不是原始数据, 而是`model_前`的输出

Reviewer: 是的，分组顺序应该是  Pipeline Parallel -> Data Parallel -> Tensor Parallel ？

Reviewer 3: 我怎么感觉分组顺序是  Pipeline Parallel -> Tensor Parallel  -> Data Parallel？

8卡先分`pipeline`, 切成`0123`和`4567`,  然后再分`tensor`,  切成`01`和`23`, 然后再分`Data`, 切成`0`和`1`

Reviewer 3: 这个应该和分组顺序关系不大,  应该和placement以及sbp的写法有关

Reviewer 2: 确认了一下，就是德澎说的这种划分

Reviewer 2: 就是说目前的代码写法下，gpu 会如何划分

Author: 我感觉没有表达错误，应该是Pipeline -> Data -> Tensor 才对, 先分Pipeline，然后每个Stage里就是前文描述的Data + Tensor的情况

Reviewer: 这里我没疑问，主要是上面 **Data Parallel + Pipeline Parallel for 2D Parallel Training on 8 GPUs** 那一节，是先 data 再 pipeline

Author: 好吧，确实有点尴尬这块儿= =，是不是要统一个写法，先写stage再data= =

> In this example, 8 GPUs will also be splitted into **2 stages**, different layers in the model will be put on different stages for pipeline parallel training. Each stage only contains a portion of the whole model, and each stage will be splitted into **2 groups**. In each stage, the model is replicated between **2 data parellel groups**, each **data parallel group** contains **2 GPUs**, the input data will be splitted into 2 parts by chunking in the batch dimension for data parallel training between **2 data parallel groups**, within each group, the weight of each layers will be splited across **2 GPUs** for tensor parallel training.

Reviewer 2: ```suggestion
In this example, 8 GPUs will also be split into **2 stages**, different layers in the model will be put on different stages for pipeline parallel training. Each stage only contains a portion of the whole model, and each stage will be split into **2 groups**. In each stage, the model will be replicated between **2 data-parallel groups**, each **data-parallel group** contains **2 GPUs**, the input data will be split into 2 parts by chunking in the batch dimension for data-parallel training between **2 data-parallel groups**, within each group, the weight of each layer will be split across **2 GPUs** for tensor parallel training.
```

Author: 已经把所有的`splitted`修改为`be split`了~

```python
from .common.train import train
...

train.dist.data_parallel_size = 2
train.dist.tensor_parallel_size = 2
train.dist.pipeline_parallel_size = 2
```

**Note:** `train.dist.data_parallel_size` will be automatically calculated by `(gpu_nums / (tensor_parallel_size * pipeline_parallel_size))` if only `train.dist.tensor_parallel_size` and `train.dist.pipeline_parallel_size` are setted, for example:
```python
from .common.train import train
...
# only set tensor_parallel_size and pipeline_parallel_size
train.dist.tensor_parallel_size = 2
train.dist.pipeline_parallel_size = 2
```
And the `data_parallel_size` will be automatically setted to `(8 / (2 * 2)) = 2`

## Update Distributed Config with Command Line

> You can also control the parallelization strategy by **command line** paremeters as follows:

Author: 已修改~

# Distributed Configuration

> In LiBai, you can try out different parallel strategy by simple changing the distributed config in [training config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/common/train.py).

Reviewer 2: ```suggestion
In LiBai, you can try out different parallel strategies by simply changing the distributed config in [training config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/common/train.py).
```

Author: 好的我改一下~

Author: 已修复~

```python
# Distributed arguments
dist=dict(
        data_parallel_size=1,
        tensor_parallel_size=1,
        pipeline_parallel_size=1,
)
```
For example, you can set `data_parallel_size=2` which will automatically split the input data into two groups for data parallel training.

## Distributed Setting Example

> Here we provide simple examples for users to understand the basic configuration of LiBai's distributed settings. LiBai's **BERT** model supports three parallelism techniques (**data parallel training**, **tensor parallel training** and **pipeline parallel training**), and here we use 1 node with 8 GPUs as an example. If you do not change any default settings, LiBai will execute **data parallel training** by default. You can try out different combinations of parallelism training techniques by updating [bert config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/bert_large_pretrain.py) as follows:

Reviewer 2: ```suggestion
Here we provide simple examples for users to understand the basic configuration of LiBai's distributed settings. LiBai's **BERT** model supports three parallelism techniques (**data parallel training**, **tensor parallel training**, and **pipeline parallel training**), and here we use 1 node with 8 GPUs as an example. If you do not change any default settings, LiBai will execute **data parallel training** by default. You can try out different combinations of parallelism training techniques by updating [bert config file](https://github.com/Oneflow-Inc/libai/blob/main/configs/bert_large_pretrain.py) as follows:
```

Author: 这儿似乎不用加逗号

```python
from .common.train import train
...

train.dist.data_parallel_size = 8
```

#### **Pure Tensor Parallel Training on 8 GPUs**

In this example, the weight of the layers in the model will be split into 8 parts for tensor parallel training on 8 GPUs.
```python
from .common.train import train
...

train.dist.tensor_parallel_size = 8
```
**Note:** Only work for models built with ``libai.layers``.

#### **Pure Pipeline Parallel Training on 8 GPUs**

In this example, 8 GPUs will be split into 8 stages, different layers of the model will be put on different stages automatically for pipeline parallel training.
```python
from .common.train import train
...

train.dist.pipeline_parallel_size = 8
```

**Note:** For models which have been configured with pipeline parallelism(e.g., BERT, GPT-2, T5 and ViT), you can simply update the distributed config to execute pipeline parallel training on them. If you need to train your own model with pipeline parallel strategy, please refer to [Write Models]() for more details about configuring your own model with pipeline parallelism.

#### **Data Parallel + Tensor Parallel for 2D Parallel Training on 8 GPUs**

> In this example, 8 GPUs will be split into **2 groups**, each group contains **4 GPUs**, and the input data will be split into 2 parts by chunking in the batch dimension for data parallel training between 2 groups. The model is replicated between **2 data parellel groups**, within each group, the weight of each layers will be splited across **4 GPUs** for tensor parallel training.

Reviewer 2: ```suggestion
In this example, 8 GPUs will be split into **2 groups**, each group contains **4 GPUs**, and the input data will be split into 2 parts by chunking in the batch dimension for data parallel training between 2 groups. The model is replicated between **2 data-parallel groups**, within each group, the weight of each layer will be split across **4 GPUs** for tensor parallel training.
```

Author: 奥这儿我fix了~等我push上来

Author: 已修复~

```python
from .common.train import train
...

train.dist.data_parallel_size = 2
train.dist.tensor_parallel_size = 4
```

> Here we provide a specific example for you to understand this, we number 8 GPUs from 0 to 7, e.g., ``[0, 1, 2, 3, 4, 5, 6, 7]``, and for ``data parallel + tensor parallel``, 8 GPUs will be split into 2 groups as ``[[0, 1, 2, 3], [4, 5, 6, 7]]``, ``GPU: [0, 1, 2, 3]`` as group-0 and ``GPU: [4, 5, 6, 7]`` as group-1. The model is replicated between group-0 and group-1. In group-0, the model will execute tensor parallel between ``GPU: [0, 1, 2, 3]``. In group-1, the model will execute tensor parallel between ``GPU: [4, 5, 6, 7]``, and each group only handle a portion of the input data for data parallel training.

Author: 这里增加了一个更加具体的例子，应该很有助于理解 @Ldpe2G

Reviewer: 好的

#### **Data Parallel + Pipeline Parallel for 2D Parallel Training on 8 GPUs**

> In this example, 8 GPUs will be split into **2 groups**, each group contains **4 GPUs**, and the input data will be split into 2 parts by chunking in the batch dimension for data parallel training. The model is replicated between **2 data parellel groups**, and each group contains **4 stages**, different layers in the model will be put on different stages automatically for pipeline parallel training.

Reviewer: 这里是先 数据分再流水

Author: OK，我改一下这部分的描述，和前后文对齐

Author: 已改~
